server/src/main/java/org/elasticsearch/plugins/RestCompatibilityPlugin.java

```diff
@@ -0,0 +1,26 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.plugins;
+
+import org.elasticsearch.Version;
+
+public interface RestCompatibilityPlugin {
+    Version minimumRestCompatibilityVersion();
+}
```

Comment:
Instead of a `Version` method, can this be a `Function` that accepts a `String` (the header) and returns `Version` such that that the function can passed into the RestController and applied as need. It will help to keep the logic encapsulated in the plugin, such that the OSS code only needs to know about the header so it can pass it to the function.

Reply:
I am actually not convinced this will be a good idea. I know that we want to as much compatible code into xpack plugin, but from Object Oriented model of RestRequest that function really belongs there. 
I made an attempt to show you how it would look if we extract this. 
The main problem is that we essentially are exposing RestRequest details (although these are still public - `hasContent` and `headers`) and pass it over to a function implemented in a xpack plugin.
Take a look and let me know

Reply:
@pgomulka I submitted a PR to your branch https://github.com/pgomulka/elasticsearch/pull/21 with some thoughts on how this can be tweaked to be a bit cleaner. 

The changes suggested (in PR):
* remove need for hasContent(), and simply look at the existence of Content-Type
* pass only the Accept and Content-Type headers through the function
* create the function as soon as possible and pass that around
* minor naming/organization such that the plugin happens to implement the functional interface, (as opposed to IS the interface)
* minor tweaking to the validation rules for the headers (since don't have hasContent())
* move all parsing logic to plugin. (some may need to go back in xContent, but for now I think this should work)

I am still +1 to functional over OO, maybe the proposed changes will help to change your mind ?  Let me know.